pkg/status/reporter.go

```diff
@@ -0,0 +1,197 @@
+	ch chan *apiv3.CalicoNodeStatus
+
+	// status holds latest version of node status resource.
+	status *apiv3.CalicoNodeStatus
+
+	// Interval and Time ticker that node status should be reported.
+	interval int
+	ticker   *time.Ticker
+
+	// populators
+	populators map[BirdConnType]map[apiv3.NodeStatusClassType]statusPopulator
+
+	// channel to indicate this reporter is not needed anymore.
+	// It should start termination process.
+	done chan struct{}
+
+	// channel to indicate this reporter is terminated.
+	term chan struct{}
+}
+
+// newReporter creates a reporter and start running a goroutine handling resource update.
+func newReporter(name string, client client.Interface, intervalInSeconds int, populators map[BirdConnType]map[apiv3.NodeStatusClassType]statusPopulator) *reporter {
+	r := &reporter{
+		name:       name,
+		client:     client,
+		ch:         make(chan *apiv3.CalicoNodeStatus, 10),
+		populators: populators,
+		done:       make(chan struct{}),
+	}
+
+	r.checkAndUpdateTicker(intervalInSeconds)
+
+	go r.run()
+	return r
+}
+
+// Check and set new ticker for the reporter.
+// Make sure stop the old one first to GC old ticker.
+func (r reporter) checkAndUpdateTicker(interval int) {
+	if r.interval == interval {
+		// no update needed.
+		return
+	}
+	if r.ticker != nil {
+		r.ticker.Stop()
+	}
+	r.interval = interval
+	r.ticker = time.NewTicker(time.Duration(interval) * time.Second)
+}
+
+// Cleanup resources owned by this reporter.
+func (r reporter) cleanup() {
+	r.ticker.Stop()
+}
+
+// KillAndWait sends done signal to reporter goroutine and wait until the
+// goroutine of the reporter terminated.
+func (r reporter) KillAndWait() {
+	r.done <- struct{}{}
+	<-r.term
+}
+
+// Called when the caller needs to send a new version of request.
+func (r reporter) RequestUpdate(request *apiv3.CalicoNodeStatus) {
+	r.ch <- request
+}
+
+// run is the main reporting loop, it loops until done.
+func (r reporter) run() {
+	runImmediately := make(chan struct{})
+
+	for {
+		select {
+		case latest := <-r.ch:
+			// Received an update of node status resource.
+			if latest.Name != r.name {
+				log.Warningf("node status reporter (%s) receive request with different name (%s), ignore it", r.name, latest.Name)
+			} else {
+				r.status = latest
+				r.checkAndUpdateTicker(latest.Spec.UpdateIntervalInSeconds)
+				// kick start node status update
+				runImmediately <- struct{}{}
+			}
+		case <-runImmediately:
+		case <-r.ticker.C:
+			if r.status == nil {
+				log.Warningf("No request found for node status reporter (%s)", r.name)
+			} else {
+				r.reportStatus()
+			}
+		case <-r.done:
+			r.cleanup()
+			r.term <- struct{}{}
+			return
+		}
+	}
+}
+
+// reportStatus queries Bird or other components and update node status resource.
+func (r reporter) reportStatus() {
+	var err error
+	// The idea here is that we either update everything successfully or we update nothing.
+
+	// Make a local copy first.
+	status := *r.status
```

Comment:
Just before we report back status, do we need to read the latest version of it? But the things is only this node is updating the resource so maybe we don't need that?

Reply:
We definitely need to handle resource conflicts, even if it should be rare. 

A user could modify the node status resource, for example.